### ISSUE TITLE
Expose the HloOpcode::kPartitionId in Jax.

### ISSUE THREAD

**Issue by ptoulme-aws** [jax-ml/jax#24721]
Hello, I think it would be very beneficial for Jax to expose the HloOpcode::kPartitionId in Jax.lax. 
The HloOpcode::kPartitionId op is the integer value of the partition running that Hlo. 
It is very useful when dynamic-slicing. 

Example: 
```
32 partitions
tensor = (32, 1024, 1024)
(1,1024,1024) = jax.lax.dynamic_slice(jax.lax.partition_id, 1024, 1024)
```

I am willing to take on this project, but any pointers would be helpful.

**Comment by justinjfu** [comment]
The general process for this would be:

1. Add a new JAX primitive (e.g. `partition_id_p = core.Primitive(...)`)
2. Register an MLIR lowering for that primitive.
```
from jax._src.interpreters import mlir

def your_lowering(...):
  ...
mlir.register_lowering(partition_id_p, your_lowering)
```
You'll also need to register an abstract eval to provide shape information to JAX.

3. Call the HLO Python bindings inside of the lowering function.
In this case, it seems like you this already exists in XLA https://openxla.org/stablehlo/spec#partition_id, so the lowering would look something along the lines of:
```
from jax._src.lib.mlir.dialects import hlo

def your_lowering(...):
   return hlo.partition_id(...)
```

**Comment by ptoulme-aws** [comment]
That is pretty cool. Let me give it a try. It is very useful op, especially in shard_maps.

**Comment by justinjfu** [comment]
Feel free to tag me on a PR or ask follow-up questions if you get stuck.

**Comment by yashk2810** [comment]
Wouldn't running `jax.lax.axis_index(axis)` inside a shard_map do the same thing for you?

**Comment by ptoulme-aws** [comment]
I see so i could within a shard_map do 

```
jax.lax.dynamic_slice(jax.lax.axis_index('model'))
```
That is not exactly the same as partition id. Take for example a mesh of Mesh(model=8, dp=4) so 32 partitions

and I have a tensor of rank (32,1024,1024)

jax.lax.dynamic_slice(jax.lax.partition_id(), :, :)
would give a (1,1024,1024) slice per rank. I am confused how to replicate that with the above 2D mesh and axis_index @yashk2810

**Comment by yashk2810** [comment]
`jax.lax.dynamic_slice(jax.lax.axis_index(('dp, 'model')))` doesn't work?

**Comment by ptoulme-aws** [comment]
@yashk2810  That did work!
I suppose we no longer need the partition id lowering then. 

```
from functools import partial

import jax
import jax.numpy as jnp

from jax.sharding import Mesh, PartitionSpec as P
from jax.experimental.shard_map import shard_map

# Create the mesh with dimensions (4, 2) and named axes ('model', 'data')
mesh = mesh = jax.make_mesh((4,2), ('model', 'data'))

# Create a random number generator and generate a tensor of size (8, 1024, 1024)
key = jax.random.PRNGKey(0)
rng_tensor = jax.random.uniform(key, (8, 1024, 1024))

# Function to apply dynamic slicing in a shard_map
@partial(shard_map, mesh=mesh, in_specs=(P(None, None, None)),
         out_specs=P(('model', 'data'), None, None))
def slice_func(args):
    # Get the mesh axis indices
    axis_index_x = jax.lax.axis_index('model')
    axis_index_y = jax.lax.axis_index('data')

    # Perform a dynamic slice (example slicing)
    # Adjust slice indices and sizes as necessary for your use case
    slice_result = jax.lax.dynamic_slice(
        args, (jax.lax.axis_index(('data', 'model')),0, 0), (1, 1024, 1024)
    )
    return slice_result

# Apply shard_map with the slice_func
jitted = jax.jit(slice_func)
result = jitted(rng_tensor)

print(jitted.lower(rng_tensor).compile().as_text())
```

HLO 
```
(8, 1024, 1024)
HloModule jit_slice_func, is_scheduled=true, entry_computation_layout={(f32[8,1024,1024]{2,1,0:T(8,128)})->f32[1,1024,1024]{2,1,0:T(8,128)}}, allow_spmd_sharding_propagation_to_parameters={true}, allow_spmd_sharding_propagation_to_output={true}, num_partitions=8

%fused_computation (param_0: f32[8,1024,1024], param_1: s32[]) -> f32[1,1024,1024] {
  %param_0 = f32[8,1024,1024]{2,1,0:T(8,128)} parameter(0)
  %param_1 = s32[]{:T(256)S(6)} parameter(1)
  %constant.26 = s32[]{:T(256)} constant(0)
  ROOT %dynamic-slice.2 = f32[1,1024,1024]{2,1,0:T(8,128)} dynamic-slice(f32[8,1024,1024]{2,1,0:T(8,128)} %param_0, s32[]{:T(256)S(6)} %param_1, s32[]{:T(256)} %constant.26, s32[]{:T(256)} %constant.26), dynamic_slice_sizes={1,1024,1024}, metadata={op_name="jit(slice_func)/jit(main)/jit(shmap_body)/dynamic_slice" source_file="<ipython-input-17-87d788ce8f04>" source_line=26}, backend_config={"flag_configs":[],"scoped_memory_configs":[],"indices_config":{"index_known_bits":[{"zeroes":"0","ones":"0","bitwidth":"32"},{"zeroes":"4294967295","ones":"0","bitwidth":"32"},{"zeroes":"4294967295","ones":"0","bitwidth":"32"}]},"used_scoped_memory_configs":[]}
}

ENTRY %main.29_spmd (param: f32[8,1024,1024]) -> f32[1,1024,1024] {
  %constant.24 = s32[]{:T(256)} constant(8)
  %constant.23 = s32[]{:T(256)} constant(0)
  %constant.22 = s32[]{:T(256)} constant(4)
  %constant.20 = u32[]{:T(256)} constant(3), metadata={op_name="jit(slice_func)/jit(main)/jit(shmap_body)/axis_index" source_file="<ipython-input-17-87d788ce8f04>" source_line=27}
  %constant.19 = u32[]{:T(256)} constant(1), metadata={op_name="jit(slice_func)/jit(main)/jit(shmap_body)/axis_index" source_file="<ipython-input-17-87d788ce8f04>" source_line=27}
  %param = f32[8,1024,1024]{2,1,0:T(8,128)} parameter(0), sharding={replicated}, metadata={op_name="args"}
  %partition-id.3 = u32[]{:T(256)} partition-id(), metadata={op_name="jit(slice_func)/jit(main)/jit(shmap_body)/axis_index" source_file="<ipython-input-17-87d788ce8f04>" source_line=27}
  %and.3 = u32[]{:T(256)S(6)} and(u32[]{:T(256)} %partition-id.3, u32[]{:T(256)} %constant.19), metadata={op_name="jit(slice_func)/jit(main)/jit(shmap_body)/axis_index" source_file="<ipython-input-17-87d788ce8f04>" source_line=27}
  %convert.3 = s32[]{:T(256)S(6)} convert(u32[]{:T(256)S(6)} %and.3), metadata={op_name="jit(slice_func)/jit(main)/jit(shmap_body)/axis_index" source_file="<ipython-input-17-87d788ce8f04>" source_line=27}
  %multiply.1 = s32[]{:T(256)S(6)} multiply(s32[]{:T(256)S(6)} %convert.3, s32[]{:T(256)} %constant.22), metadata={op_name="jit(slice_func)/jit(main)/jit(shmap_body)/mul" source_file="<ipython-input-17-87d788ce8f04>" source_line=27}
  %shift-right-logical.1 = u32[]{:T(256)S(6)} shift-right-logical(u32[]{:T(256)} %partition-id.3, u32[]{:T(256)} %constant.19), metadata={op_name="jit(slice_func)/jit(main)/jit(shmap_body)/axis_index" source_file="<ipython-input-17-87d788ce8f04>" source_line=27}
  %and.2 = u32[]{:T(256)S(6)} and(u32[]{:T(256)S(6)} %shift-right-logical.1, u32[]{:T(256)} %constant.20), metadata={op_name="jit(slice_func)/jit(main)/jit(shmap_body)/axis_index" source_file="<ipython-input-17-87d788ce8f04>" source_line=27}
  %convert.2 = s32[]{:T(256)S(6)} convert(u32[]{:T(256)S(6)} %and.2), metadata={op_name="jit(slice_func)/jit(main)/jit(shmap_body)/axis_index" source_file="<ipython-input-17-87d788ce8f04>" source_line=27}
  %add.2 = s32[]{:T(256)S(6)} add(s32[]{:T(256)S(6)} %convert.2, s32[]{:T(256)S(6)} %multiply.1), metadata={op_name="jit(slice_func)/jit(main)/jit(shmap_body)/add" source_file="<ipython-input-17-87d788ce8f04>" source_line=27}
  %add.3 = s32[]{:T(256)S(6)} add(s32[]{:T(256)S(6)} %add.2, s32[]{:T(256)} %constant.24), metadata={op_name="jit(slice_func)/jit(main)/jit(shmap_body)/add" source_file="<ipython-input-17-87d788ce8f04>" source_line=26}
  %compare.1 = pred[]{:T(1024)S(6)} compare(s32[]{:T(256)S(6)} %add.2, s32[]{:T(256)} %constant.23), direction=LT, metadata={op_name="jit(slice_func)/jit(main)/jit(shmap_body)/lt" source_file="<ipython-input-17-87d788ce8f04>" source_line=26}
  %select.1 = s32[]{:T(256)S(6)} select(pred[]{:T(1024)S(6)} %compare.1, s32[]{:T(256)S(6)} %add.3, s32[]{:T(256)S(6)} %add.2), metadata={op_name="jit(slice_func)/jit(main)/jit(shmap_body)/select_n" source_file="<ipython-input-17-87d788ce8f04>" source_line=26}
  ROOT %fusion = f32[1,1024,1024]{2,1,0:T(8,128)} fusion(f32[8,1024,1024]{2,1,0:T(8,128)} %param, s32[]{:T(256)S(6)} %select.1), kind=kLoop, calls=%fused_computation, metadata={op_name="jit(slice_func)/jit(main)/jit(shmap_body)/dynamic_slice" source_file="<ipython-input-17-87d788ce8f04>" source_line=26}, backend_config={"flag_configs":[],"window_config":{"kernel_window_bounds":[],"output_window_bounds":["1","32","8"],"input_window_bounds":[],"estimated_cycles":"19784","iteration_bounds":["1","4","1"]},"scoped_memory_configs":[],"used_scoped_memory_configs":[{"memory_space":"1","offset":"0","size":"4194304"}],"retry_config":{"retry_count":"0"}}
}
```
